doc/threat_model/Application_Loader.md

```diff
@@ -0,0 +1,88 @@
+Application Loader
+==================
+
+## What is an Application Loader?
+
+The term "application loader" refers to the mechanism used to add Tock
+applications to a Tock system. It can take several forms; here are a few
+examples:
+
+1. Tockloader is an application loader that runs on a host system. It uses
+   programming interfaces to manipulate applications on the Tock system's
+   nonvolatile storage.
+
+1. Some build systems combine the kernel and apps at build time into a single,
+   monolithic image. This monolithic image is then deployed using a programming
+   tool.
+
+1. A kernel-assisted installer may be a Tock capsule that receives an
+   application over USB and writes it into flash.
+
+## Why Must We Trust It?
+
+The application loader has the ability to erase, rewrite, and sometimes corrupt
+applications. As a result, the application loader must be trusted to provide
+confidentiality, availability, and sometimes integrity guarantees to
+applications. For example, the application loader must not modify, erase, or
+exfiltrate applications other than the application(s) it was asked to operate
+on.
+
+Tock kernels that require all applications to be signed do not need to trust the
+application loader for application integrity, as that is done by validating the
+signature instead. However, the application loader still needs to be trusted for
+availability guarantees, as it can still modify applications and make the
+signature check fail. Tock kernels that do not require signed applications must
+trust the application loader to not maliciously modify applications.
+
+To protect the kernel's confidentiality, integrity, and availability the
+application loader must not modify, erase, or exfiltrate kernel data. On most
+boards, the application loader must be trusted to not modify, erase, or
+exfiltrate kernel data. However, Tock boards may use other mechanisms to protect
+the kernel without trusting the application loader. For example, a board with
+access-control hardware between its flash storage and the application loader may
+use that hardware to protect the kernel's data without trusting the application
+loader.
+
+## TBF Verification Requirement
+
+The application loader is required to confirm that the TBF's `total_size` field
+is correct (as specified in the [Tock Binary
+Format](../TockBinaryFormat.md#tbf-header-base)) before deploying an
+application. This is to prevent the newly-deployed application from executing
+the following attacks:
+
+1. Specifying an incorrect `total_size`, preventing the kernel from finding the
+   remaining application(s), which prevents the remaining application(s) from
+   launching (impacting availability).
+
+1. Specifying a too-large `total_size` that includes the subsequent
+   application(s) image(s), allowing the malicious application to read the
+   images(s) (impacting availability and confidentiality).
+
```

Comment:
Hmm...this may not be enough of a requirement. Since we don't require that there be any particular sentinel at the end of the valid apps in flash, it is difficult (in light of an adversary) to tell when we are at the end of the list of applications.

Consider a list of say four applications, and the app loader wants to corrupt the third app's TBF header. It can set the total_size in the header correctly, but if it sets the other initial fields incorrectly the header will fail to parse and the kernel will not load the fourth app.

At the very least we need the version half-word to be correct as well.

Reply:
I added wording indicating that the length field check is version-dependent.